### PR TITLE
helm 3.0.0

### PR DESCRIPTION
**Pull request by mikesplain**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
We may want to consider holding this considering helm 2 is still going to be in play for many users, e.g. https://github.com/Homebrew/homebrew-core/pull/46671#issuecomment-553433509

### CODE REVIEW

**Comment by mikesplain** [comment]
Also opened #46709, since we should probably do that instead but will leave this open until I hear from others.

**Comment by cagataygurturk** [comment]
As we discussed on https://github.com/Homebrew/homebrew-core/pull/46671#issuecomment-553503691 this may break some users‘ workflow since it is a breaking change and there won‘t be a way to install Helm 2 also, at least through Homebrew. Some plugins (e.gy Quay Registry plugin) are not compatible with Helm 3 and anyways some migration efforts are needed to make an existing cluster Helm 3 compatible. I don‘t think every user is ready for that.

I don‘t know how such a decision is made though. I would keep this Formula with 2.x and kickoff a new one with 3.x.

**Comment by fxcoudert** [comment]
For context, Homebrew's rules for versioned formulas are here: https://docs.brew.sh/Versions
In the context of helm, what should happen is:

- `helm` formula provides the latest stable version, i.e. 3.0.0
- If the Helm 2 branch continues to be maintained upstream (i.e. there will be new releases in the future, with security fixes at least), then we can create a `helm@2` formula that will track the helm 2 branch.

There are downsides, which have been cited: people will by default follow the major upgrade path, but:
- they can fix it with a simple line
- Homebrew's users should not be surprised, as this is the documented behavior
- other paths lead to a lot of pain later (cough `openssl@1.1` cough)

**Comment by cagataygurturk** [comment]
`kubernetes-helm@2` is created here: https://github.com/Homebrew/homebrew-core/pull/46721

**Comment by bacongobbler** [comment]
`glide` should be removed from the dependency tree. Helm 3 uses Go modules instead.

**Comment by bacongobbler** [comment]
The `--client` flag can be removed as well from the version check. Helm 3 will silently accept it, but I don't see a need to keep using that going forward.

(sorry, have to comment here as the lines referenced weren't changed in the PR)

**Comment by bacongobbler** [comment]
> dir = buildpath/"src/k8s.io/helm"

Should be `dir = buildpath/"src/helm.sh/helm"`, as the go paths have changes as well. https://helm.sh/docs/faq/#go-import-path-changes

**Comment by mikesplain** [comment]
Changes made, thanks @bacongobbler and all.

**Comment by scottrigby** [comment]
This and the @2 PR look good 👍 so don't want to bikeshed here, I'm just not sure where else to ask. Is there an option to discuss renaming this formula? @bacongobbler https://github.com/Homebrew/homebrew-core/pull/45789#issuecomment-546190893 and @cagataygurturk https://github.com/Homebrew/homebrew-core/pull/46671#issuecomment-553503691 noted Helm is no longer a sub-project of Kubernetes.

@bacongobbler suggested `cncf-helm` (since it's now a CNCF project) because there was some memory of perhaps a namespacing issue with helm years ago, but perhaps that's no longer an issue now that Helm is so widely used - could we rename to just `helm` and `helm@2`?

Is there a better place to follow-up discussing the possibility of renaming this `kubernetes-helm` and `kubernetes-helm@2` formulas, and/or what does that process look like? Searching the Homebrew docs I haven't found it yet.

**Comment by fxcoudert** [comment]
- `helmfile` test is failing:

```
==> /usr/local/opt/kubernetes-helm/bin/helm init --client-only
Error: unknown flag: --client-only
```

- `helmsman` test is failing:

```
==> /usr/local/Cellar/helmsman/1.13.0/bin/helmsman version
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
main.init.0()
	/private/tmp/helmsman-20191030-53227-1oqxg9t/src/github.com/Praqma/helmsman/init.go:103 +0x1b70
```

- Versioned formula should be added in the same pull request, if you want the two to go in together (https://github.com/Homebrew/homebrew-core/pull/46721)

**Comment by fxcoudert** [comment]
> could we rename to just helm and helm@2?

Yes, I think it's reasonable indeed.

**Comment by cagataygurturk** [comment]
@mikesplain I have closed @2 PR. Can you add it here to this PR? Thanks.

**Comment by mattfarina** [comment]
FYI, Helm v2 will be supported for 1 year following the release of Helm v3.

**Comment by mikesplain** [comment]
Updated the PR based on the above discussion.  Currently I've set the rename from `kubernetes-helm` -> `helm`.
https://github.com/Homebrew/homebrew-core/blob/d4a6381a9621474f4859760235cab232706278b3/formula_renames.json#L109
Should we consider setting the formula rename file to `helm@2` and allow users to install `helm` to get helm 3? cc: @mattfarina, @cagataygurturk @fxcoudert @scottrigby @bacongobbler

**Comment by anthonydahanne** [comment]
@mikesplain aren't you supposed to create several commits (one for `helm` update to 3, one for `helm@2`, and even another one for `kubernetes-helm` -> `helm`) ?

**Comment by fxcoudert** [comment]
> Currently I've set the rename from kubernetes-helm -> helm

Yes.

> create several commits

Yes please. One for the formula update and rename (commit message: `helm 3.0.0`) and one for the older version (`helm@2 2.x.y (new formula)`)

**Comment by mikesplain** [comment]
Thanks all, changed as per request.

**Comment by SMillerDev** [comment]
Can you also make an entry in the Aliases directory pointing to helm.rb called helm@3?

**Comment by mikesplain** [comment]
Thanks @SMillerDev done.

File: Formula/helm@2.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+class HelmAT2 < Formula
+  desc "The Kubernetes package manager"
+  homepage "https://helm.sh/"
+  url "https://github.com/helm/helm.git",
+      :tag      => "v2.16.1",
+      :revision => "bbdfe5e7803a12bbdf97e94cd847859890cf4050"
+  head "https://github.com/helm/helm.git"
```
Comment (chenrui333) [inline Formula/helm@2.rb]: This line can be removed per audit log

08:42:57 ==> brew audit helm@2 --online --new-formula
08:44:28 ==> FAILED
08:44:28 Error: 1 problem in 1 formula detected
08:44:28 helm@2:
08:44:28   * Formulae should not have a `HEAD` spec

Comment (mikesplain) [reply Formula/helm@2.rb]: Done, thanks @chenrui333

**Comment by SMillerDev** [comment]
Looks good to me, thanks @mikesplain and others that chimed in. Without people like you it would be impossible to maintain homebrew to the high standards people have come to expect of us. You rock 🎉

**Pull request closed without merging**